boot/assets.go

```diff
@@ -632,3 +632,117 @@ func (o *TrustedAssetsUpdateObserver) Canceled() error {
 	// - reseal with a given state of modeenv
 	return nil
 }
+
+func observeSuccessfulBootWithAssetsForBootloader(m *Modeenv, root string, opts *bootloader.Options) (drop []*trackedAsset, err error) {
+	trustedAssetsMap := &m.CurrentTrustedBootAssets
+	otherTrustedAssetsMap := m.CurrentTrustedRecoveryBootAssets
+	whichBootloader := "run mode"
+	if opts != nil && opts.Role == bootloader.RoleRecovery {
+		trustedAssetsMap = &m.CurrentTrustedRecoveryBootAssets
+		otherTrustedAssetsMap = m.CurrentTrustedBootAssets
+		whichBootloader = "recovery"
+	}
+
+	if len(*trustedAssetsMap) == 0 {
+		// bootloader may have trusted assets, but we are not tracking
+		// any for the boot process
+		return nil, nil
+	}
+
+	// let's find the bootloader first
+	bl, trustedAssets, err := findMaybeTrustedAssetsBootloader(root, opts)
+	if err != nil {
+		return nil, err
+	}
+	if len(trustedAssets) == 0 {
+		// not a trusted assets bootloader, nothing to do
+		return nil, nil
+	}
+
+	cache := newTrustedAssetsCache(dirs.SnapBootAssetsDir)
+	for _, trustedAsset := range trustedAssets {
+		assetName := filepath.Base(trustedAsset)
+
+		// find the hash of the file on disk
+		assetHash, err := cache.fileHash(filepath.Join(root, trustedAsset))
+		if err != nil && !os.IsNotExist(err) {
+			return nil, fmt.Errorf("cannot calculate the digest of existing trusted asset: %v", err)
+		}
+		if assetHash == "" {
+			// no trusted asset on disk, but we booted nonetheless,
+			// at least log something
+			logger.Noticef("system booted without %v bootloader trusted asset %q", whichBootloader, trustedAsset)
+			// given that asset names cannot be reused, clear the
+			// boot assets map for the current bootloader
+			delete(*trustedAssetsMap, assetName)
+			continue
+		}
+
+		// this is what we booted with
+		bootedWith := []string{assetHash}
+		// one of these was expected during boot
+		hashList := (*trustedAssetsMap)[assetName]
+
+		assetFound := false
+		// find out if anything needs to be dropped
+		for _, hash := range hashList {
+			if hash == assetHash {
+				assetFound = true
+				continue
+			}
+			if !isAssetHashTrackedInMap(otherTrustedAssetsMap, assetName, hash) {
+				// asset can be dropped
+				drop = append(drop, &trackedAsset{
+					blName: bl.Name(),
+					name:   assetName,
+					hash:   hash,
+				})
+			}
+		}
+
+		if !assetFound {
+			// unexpected, we have booted with an asset whose hash
+			// is not listed among the ones we expect
+
+			// TODO:UC20: try to restore the asset from cache
+			return nil, fmt.Errorf("system booted with unexpected %v bootloader asset %q hash %v", whichBootloader, trustedAsset, assetHash)
+		}
+
+		// update the list of what we booted with
+		(*trustedAssetsMap)[assetName] = bootedWith
+
+	}
+	return drop, nil
+}
+
+// observeSuccessfulBootWithAssets makes not of a state of the trusted boot
```

Comment:
s/makes not of a state/observes the state/ ?

Reply:
uhh, some mental relapse on my side